feature blog: contextual logging update for 1.29 (#43579)
kubernetes/website: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md

`pod` and `node` already get logged as parameters in individual log calls in `kube-scheduler` code.
Once contextual logging is supported by more packages outside of `kube-scheduler`, 
they will also be visible there (for example, client-go). Once it is GA,
log calls can be simplified to not repeat those values.

In kube-controller-manager, `WithName` is used to add the user-visible controller name to log output, 
for example:

> I1113 08:43:29.284360   87141 graph_builder.go:285] "garbage controller monitor not synced: no monitors" **logger="garbage-collector-controller"**

The `logger=”garbage-collector-controller”` was added by the kube-controller-manager core 
when instantiating that controller and appears in all log entries of it - at least as long as the code
that it calls supports contextual logging. Further work is needed to convert shared packages like client-go.

## Performance impact

Supporting contextual logging in a package, i.e. accepting a logger from a caller, is cheap. 
No performance impact was observed for kube-scheduler. As noted above, 
adding `WithName` and `WithValues` needs to be done more carefully.

> In Kubernetes 1.29, enabling contextual logging at production verbosity (-v3 or lower) 

Reviewer: ```suggestion
In Kubernetes 1.29, enabling contextual logging at production verbosity (`-v3` or lower) 
```

Author: Done.

> caused no measurable slowdown for `kube-scheduler` and is not expected for `kube-controller-manager` either. 

Reviewer: ```suggestion
caused no measurable slowdown for the `kube-scheduler` and is not expected for the
`kube-controller-manager` either. 
```

Author: Done. Thanks for the catch!

---
layout: blog
title: "Contextual logging in Kubernetes 1.29: Better troubleshooting and enhanced logging"
slug: contextual-logging-in-kubernetes-1-29
date: 2023-12-06
---

**Author**: [Mengjiao Liu](https://github.com/mengjiao-liu/) (DaoCloud), [Patrick Ohly](https://github.com/pohly) (Intel)

On behalf of the [Structured Logging Working Group](https://github.com/kubernetes/community/blob/master/wg-structured-logging/README.md) 
and [SIG Instrumentation](https://github.com/kubernetes/community/tree/master/sig-instrumentation#readme), 
we are pleased to announce that the contextual logging feature
introduced in Kubernetes v1.24 has now been successfully migrated to

> 2 components(kube-scheduler and kube-controller-manager)

Reviewer 2: ```suggestion
two components (kube-scheduler and kube-controller-manager)
```

Author: Done.

  That output can also be more verbose by default because it will not get shown for successful tests.
  Tests can be run in parallel without interleaving their output.

One of the design decisions for contextual logging was to allow attaching a logger as value to a `context.Context`.
Since the logger encapsulates all aspects of the intended logging for the call, 
it is *part* of the context and not just *using* it. A practical advantage is that many APIs 
already have a `ctx` parameter or adding one has additional advantages, like being able to
get rid of `context.TODO()` calls inside the functions.

## How to use it

The contextual logging feature is alpha starting from Kubernetes v1.24,
so it requires the ContextualLogging feature gate to be enabled.
If you want to test the feature while it is alpha, you need to enable this feature gate
on the `kube-controller-manager` and the `kube-scheduler`.

For the `kube-scheduler`, there is one thing to note, in addition to enabling 
the `ContextualLogging` feature gate, instrumentation also depends on log verbosity.
To avoid slowing down the scheduler with the logging instrumentation for contextual logging added for 1.29, 

> it is important to be careful about the logging verbosity when using contextual logging. 
> 
> Here are the specific details:

Reviewer 2: ```suggestion
it was important to be choose carefully when to add additional information:
```

> - At -v3 or lower, only WithValues("pod") is used once per scheduling cycle. 

Reviewer 2: ```suggestion
- At -v3 or lower, only `WithValues("pod")` is used once per scheduling cycle. 
```

  This has the intended effect that all log messages for the cycle include the pod information. 
  Once contextual logging is GA, "pod" key/value pairs can be removed from all log calls.

> - At -v4 or higher, richer log entries get produced where WithValues is also used for the node (when applicable)

Reviewer 2: ```suggestion
- At -v4 or higher, richer log entries get produced where `WithValues` is also used for the node (when applicable)
```

>   and WithName is used for the current operation and plugin.

Reviewer 2: ```suggestion
  and `WithName` is used for the current operation and plugin.
```
